msg/templates/urtps/microRTPS_timesync.h.em

```diff
@@ -0,0 +1,212 @@
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/*!
+ * @@file microRTPS_timesync.h
+ * @@brief Adds time sync for the microRTPS bridge
+ * @@author Nuno Marques <nuno.marques@@dronesolutions.io>
+ * @@author Julian Kent <julian@@auterion.com>
+ */
+
+#pragma once
+
+#include <atomic>
+#include <functional>
+#include <thread>
+
+@[if ros2_distro]@
+#include "Timesync_Publisher.h"
+#include "Timesync_Subscriber.h"
+@[else]@
+#include "timesync_Publisher.h"
+#include "timesync_Subscriber.h"
+@[end if]@
+
+static constexpr double ALPHA_INITIAL = 0.05;
+static constexpr double ALPHA_FINAL = 0.003;
+static constexpr double BETA_INITIAL = 0.05;
+static constexpr double BETA_FINAL = 0.003;
+static constexpr int WINDOW_SIZE = 500;
+static constexpr int64_t UNKNOWN = 0;
+static constexpr int64_t TRIGGER_RESET_THRESHOLD_NS = 100ll * 1000ll * 1000ll;
+static constexpr int REQUEST_RESET_COUNTER_THRESHOLD = 5;
+
+@# Sets the timesync DDS type according to the FastRTPS and ROS2 version
+@[if 1.5 <= fastrtpsgen_version <= 1.7]@
+@[    if ros2_distro]@
+using timesync_msg_t = @(package)::msg::dds_::Timesync_;
+@[    else]@
+using timesync_msg_t = timesync_;
+@[    end if]@
+@[else]@
+@[    if ros2_distro]@
+using timesync_msg_t = @(package)::msg::Timesync;
+@[    else]@
+using timesync_msg_t = timesync;
+@[    end if]@
+@[end if]@
+
+class TimeSync {
+public:
+	TimeSync();
+	virtual ~TimeSync();
+
+	/**
+	 * @@brief Starts the timesync publishing thread
+	 * @@param[in] pub The timesync publisher entity to use
+	 */
+@[if ros2_distro]@
+	void start(const Timesync_Publisher* pub);
```

Comment:
Do the same thing with the Timesync_Publisher / Timesync_Subscriber with the `using =` and it will move basically all of the extra defines out of the code, into a single config at the top.

Reply:
Sure

Reply:
Done